rfcs/62-editor-data-model.md

```diff
@@ -0,0 +1,653 @@
+}
+```
+
+The `EnumType` component defines the set of entries of an enum or flags, the storage type, and a boolean indicating whether the type is an enum or a flags.
+
+```rust
+#[derive(Component)]
+struct EnumType {
+    entries: Vec<EnumEntry>,
+    storage_type: IntegralType,
+    is_flags: bool,
+}
+```
+
+#### Arrays
+
+Arrays are dynamically-sized homogenous collections of _elements_. The `ArrayType` defines the element type, which can be any valid game type.
+
+```rust
+#[derive(Component)]
+struct ArrayType {
+    elem_type: AnyType,
+}
+```
+
+Array values store their elements contiguously in the byte storage (see [Game objects](#game-objects)), without padding between elements.
+
+#### Dictionaries
+
+Dictionaries are dynamically-sized homogenous collections mapping _keys_ to _values_, where each key is unique in the instance. The key type is always a string (`Name` type). The `DictType` defines the value type.
+
+```rust
+#[derive(Component)]
+struct DictType {
+    value_type: AnyType,
+}
+```
+
+Dictionary values store their (key, value) pairs contiguously in the byte storage (see [Game objects](#game-objects)), key first followed by value, without padding.
+
+#### Structs
+
+A _struct_ is an heterogenous aggregation of other types (like a `struct` in Rust). The `StructType` component contains the _properties_ of the struct.
+
+```rust
+#[derive(Component)]
+struct StructType {
+    properties: Vec<Property>, // sorted by `byte_offset`
+}
+```
+
+Properties themselves are defined by a name and the type of their value. The type also stores the byte offset of that property from the beginning of a struct instance, which allows packing property values for a struct into a single byte stream, and enables various optimizations regarding diff'ing/patching and serialization. Modifying any of the `name`, `value_type`, or `byte_offset` constitute a type change, which mandates migrating all existing instances.
+
+```rust
+struct Property {
+    name: Name,
+    value_type: AnyType,
+    byte_offset: u32,
+    validate: Option<Box<dyn Validate>>,
+}
+```
+
+The _type layout_ refers to the collection of property offsets and types of an `StructType`, which together define:
+
+- the order in which the properties are listed in the struct type (the properties are sorted in increasing byte offset order), and any instance value stored in a `GameObject`.
+- the size in bytes of each property value, based on its type.
+- eventual gaps between properties, known as _padding bytes_, depending on their offsets and sizes.
+
+Unlike Rust, the Editor data model guarantees that **padding bytes are zeroed**. This property enables efficient instance comparison via `memcmp()`-like byte-level comparison, without the need to rely on individual properties inspection nor any equality operator (`PartialEq` and `Eq` in Rust). It also enables efficient copy and serialization via direct memory reads and writes, instead of relying on getter/setter methods.

```

Comment:
I wonder if this will affect the memory layout for nested components. Like a `Transform` component compiled on the game client may have different layout from a `Transform` component compiled on Bevy Editor, since Bevy Editor will be a standalone executable. 

I'm no expert, but since `Transform` doesn't have `#[repr(C)]` direct memory operation on struct properties wont be reliable.

Reply:
I'm not sure where I lost you in my explanation 🤣 but happy to try and clarify. The Editor will never ever attempt to take data from a Game `Transform` and store it into an Editor `Transform`. The former exists only in the Game process, and in the Editor is represented by a `GameType` / `GameObject`, _not_ a `Transform`. And Editor's `Transform` are only used for, say, rendering the UI of the Editor itself; there's no relation with the Game data being edited.

Reply:
You may want to think of `GameObject` has a `DynamicStruct` if that makes more sense: https://docs.rs/bevy/latest/bevy/reflect/struct.DynamicStruct.html. This might be how we actually implement those.